File: Single Number II/Single_Number_II.cpp
// 这道题目剑指offer上应该是给错了，因为memo中存储的是数字的补码，不能直接转换为数字的!!!!!!
// 同样，如果这道题目出现在面试中，应该考虑到各种边界条件
// 这道题目中的算法是可以扩展到N个重复数字的

// Runtime: 12 ms, faster than 93.36% of C++ online submissions for Single Number II.
// Memory Usage: 9.6 MB, less than 78.65% of C++ online submissions for Single Number II.

class Solution
{
public:
	int singleNumber(vector<int>& nums)
	{
		vector<int> memo(32, 0);

		long temp = 1;
		for (int index = 0; index < 32; index++, temp <<= 1)
		{
			for (auto number : nums)
			{
				memo[index] += ((temp & number) != 0);
			}
			memo[index] %= 3;
		}

		// 如果符号位为1，则需要将补码转换为原码
        if (memo[31] == 1)
        {
            bool found = false;
            for (int index = 0; index <= 30; index++)
            {
                if (memo[index] == 1)
                {
                    found = true;
                    memo[index] = 0;
                    while (--index >= 0)
                        memo[index] = 1;
                    break;
                }
            }
            if (!found)
                return INT_MIN;

            for (int i = 0; i <= 30; i++)
                memo[i] = (memo[i] == 0);
        }

		// 将补码转换为数字
		int res = 0;
		for (int index = 30; index >= 0; index--)
		{
			res <<= 1;
			res += memo[index];
		}
		return memo[31] == 0 ? res : -res;
	}
};

File: Single Number II/Single_Number_II.py
# Runtime: 140 ms, faster than 7.26% of Python3 online submissions for Single Number II.
# Memory Usage: 14.1 MB, less than 68.49% of Python3 online submissions for Single Number II.

class Solution:
    def singleNumber(self, nums) -> int:
        
        memo = [0 for i in range(32)]
        temp = 1
        
        for index in range(32):
            for number in nums:
                memo[index] += ((temp & number) >= 1)
            memo[index] %= 3

            temp <<= 1
        
        if memo[31] is 1:
            found = False
            for index in range(31):
                if memo[index] is 1:
                    found = True
                    memo[index] = 0
                    index -= 1
                    while(index >= 0):
                        memo[index] = 1
                        index -= 1
                    break
            if not found:
                return -(2 ** 31)
            for i in range(31):
                memo[i] = int(memo[i] == 0)
        
        res = 0
        for index in range(30, -1, -1):
            res <<= 1
            res += memo[index]
        
        return res if memo[31] == 0 else -res